codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedCollectionGenerator.kt

```diff
@@ -117,16 +117,74 @@ class UnconstrainedCollectionGenerator(
+                     * indirectly constrained:
+                     * ```smithy
+                     * @length(min: 1, max: 100)
+                     * list ItemList {
+                     *     member: Item
+                     * }
+                     * list Item {
+                     *     member: ItemName
+                     * }
+                     * @length(min: 1, max: 100)
+                     * string ItemName
+                     * ```
+                     *
+                     * For this model, two `pub(crate)` types are generated for the `Item` shape:
+                     * - `ItemUnconstrained`: represents the non-validated version
+                     * - `ItemConstrained`: represents the validated version
+                     *
+                     * Similarly, for `ItemList`:
+                     * - `ItemListUnconstrained`: a `pub(crate)` type representing the non-validated version
+                     * - `ItemList`: the publicly exposed validated version
+                     *
+                     * A `TryFrom` implementation must be generated to convert from `ItemListUnconstrained` to `ItemList`.
+                     * Since the final type exposed to the user is `struct ItemList(Vec<Vec<ItemName>>)`, the conversion
+                     * process involves two steps:
+                     * 1. Converting each element of the vector from `ItemUnconstrained` to `ItemConstrained` to validate
+                     *    constraints
+                     * 2. Converting the resulting `Vec<ItemConstrained>` to `Vec<Vec<ItemName>>`
+                     */
+                    val constrainedValueTypeIsNotFinalType =
+                        resolvesToNonPublicConstrainedValueType && shape.isDirectlyConstrained(symbolProvider)
+
+                    val finalType =
+                        if (constrainedValueTypeIsNotFinalType) {
+                            constrainedShapeSymbolProvider.toSymbol(shape.member)
+                        } else {
+                            constrainedMemberSymbol
+                        }
+
                     val constrainValueWritable =
                         writable {
                             conditionalBlock("inner.map(|inner| ", ").transpose()", constrainedMemberSymbol.isOptional()) {
-                                rust("inner.try_into().map_err(|inner_violation| (idx, inner_violation))")
+                                rustTemplate(
+                                    """
+                                    inner.try_into()
+                                        #{FinalMapping}
+                                        .map_err(|inner_violation| (idx, inner_violation))
+                                    """,
+                                    "FinalMapping" to
+                                        writable {
+                                            if (constrainedValueTypeIsNotFinalType) {
```

Comment:
I benchmarked performing this in separate phases, as done in `UnconstrainedMapGenerator`, but found it to be about 5ns slower on a vector of 4 string values. Therefore, I decided to keep this version instead of adding a second pass.